GC Tombstone: Add option for apps to pass to disable GC enforcement for old files

markfields opened the pull request:
## Description

In discussions with our internal partner teams, as they consider rolling out GC via the Tombstone feature, they want a way to cut off old documents from exposure to failure due to GC (Tombstones failing to load, or GC Sweep deleting objects).  This is because some GC-impacting bugs in partner code were forward-fixed only, without leaving repair code in place - so documents created before such a bug was fixed would be subject to being corrupted by GC (e.g. due to invalid internal document structure), if they were opened after GC was enabled.

This PR adds an otherwise undocumented ContainerRuntime Option `gcEnforcementMinCreateContainerRuntimeVersion`, which is to be compared with the value of the persisted metadata property `createContainerRuntimeVersion` to determine whether a document was created recently enough to be subjected to enforcement of GC.

Some clarifications:

* If `gcEnforcementMinCreateContainerRuntimeVersion` is omitted (the default case since it's undocumented) or malformed, GC enforcement is determined based on the existing configuration parameters.  Only if it's present and strictly `>` the file's `createContainerRuntimeVersion` will GC enforcement be disabled.
* If `createContainerRuntimeVersion` is missing from a file, it's considered older than any valid value of `gcEnforcementMinCreateContainerRuntimeVersion` such that if the option is set, GC enforcement will be disabled.

## Reviewer Guidance

There are a few todos (`//*` comments) but the basic structure is there.

I'd love special attention given to the gcHelper tests that cover the version comparison stuff.  What other test cases should I put in there?

markfields committed:
Add gcContainerGeneration and tests illustrating its usage

markfields committed:
prettier fix whitespace

markfields committed:
Redo prettier stuff, and start to move to new option

markfields committed:
npm run lint:fix

markfields committed:
api md

markfields committed:
Fix tests, rename stuff

markfields committed:
Remove comment

markfields committed:
Rename

markfields commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,28 @@ describe("Garbage Collection Tests", () => {
 				assert(gc.gcEnabled, "gcEnabled incorrect");
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value, sweepEnabled false", () => {
```
This illustrates how GC enforcement would be disabled for an old document.  It's using the `sweepEnabled` setting but would also apply to tombstone failure.

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:
```diff
@@ -602,6 +623,51 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
+				const tombstoneSuccessResponse = await containerRuntime_resolveHandle(container, {
+					url: unreferencedId,
+				});
+				assert.equal(
```
The tombstone errors will still be logged right?

markfields replied:
Yes.  But we'll want to log something about this option and the persisted value so we can filter it out.

agarwal-navin commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,28 @@ describe("Garbage Collection Tests", () => {
+			it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value, sweepEnabled false", () => {
+				gc = createGcWithPrivateMembers(
+					{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
+					{ gcEnforcementMinCreateContainerRuntimeVersion: "2.0.0" },
```
gcEnforcementMinCreateContainerRuntimeVersion is too wordy IMO :)

markfields replied:
Open to suggestions.

I like it because it's ugly.  It's undocumented so I went for highest degree of clarity with what it does (without writing a complete sentence)

markfields committed:
Add some test cases

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,28 @@ describe("Garbage Collection Tests", () => {
+			it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value, sweepEnabled false", () => {
+				gc = createGcWithPrivateMembers(
+					{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
+					{ gcEnforcementMinCreateContainerRuntimeVersion: "2.0.0" },
```
So the biggest problem with using string values here is either you have to write your own semver comparison algorithm, which is going to be challenging, or you're going to have to import a semver comparison algorithm which will cause the bundle size to increase.

markfields commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			describe("gcEnforcementMinCreateContainerRuntimeVersion overriding sweepEnabled true for some existing containers", () => {
+				it("gcEnforcementMinCreateContainerRuntimeVersion later than persisted value: sweepEnabled false", () => {
```
This illustrates how GC enforcement would be disabled for an old document.  It's using the `sweepEnabled` setting but would also apply to tombstone failure.

markfields commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
 				assert(gc.gcEnabled, "gcEnabled incorrect");
 				assert.equal(gc.latestSummaryGCVersion, 1, "latestSummaryGCVersion incorrect");
 			});
+			describe("gcEnforcementMinCreateContainerRuntimeVersion overriding sweepEnabled true for some existing containers", () => {
```
https://github.com/microsoft/FluidFramework/pull/13985/files/9fd0e174f7105671eacb0f1fe86efd5cee9bc283#r1096322059

Highlighting Navin's comment about the name being wordy - I like it but open to other suggestions :)

markfields commented on packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:
```diff
@@ -230,6 +231,44 @@ describe("Garbage Collection Tests", () => {
+				});
+				it("gcEnforcementMinCreateContainerRuntimeVersion earlier than persisted value: sweepEnabled true", () => {
+					gc = createGcWithPrivateMembers(
+						{ createContainerRuntimeVersion: "1.2.3", sweepEnabled: true },
```
Tyler [says](https://github.com/microsoft/FluidFramework/pull/13985/files/9fd0e174f7105671eacb0f1fe86efd5cee9bc283#r1096322964):

> So the biggest problem with using string values here is either you have to write your own semver comparison algorithm, which is going to be challenging as we are doing 2.0.0-internal.x.y.z and you will need to differentiate between dev versions and the like, or you're going to have to import a semver comparison algorithm which will cause the bundle size to increase.

markfields committed:
Expand/simplify test cases in gc.spec.ts

markfields committed:
Add boolean ContainerRuntime.disableGcTombstoneEnforcement

This will be computed in the constructor based on metadata and containerRuntimeOptions

markfields committed:
Test baseline tweak

markfields committed:
Move tests from gc.spec.ts to gcHelpers.spec.ts due to code layering

markfields committed:
Implement version compare and update tests. They pass!

markfields committed:
Update test case

markfields committed:
lint fixes

markfields committed:
lint fix

markfields commented:
Did #14050 instead
